cmd/aadsync/main.go

```diff
@@ -0,0 +1,200 @@
+package main
+
+import (
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac"
+	"github.com/sirupsen/logrus"
+
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openshift/api/user/v1"
+	"github.com/openshift/openshift-azure/test/clients/openshift"
+)
+
+// https://docs.microsoft.com/en-us/graph/api/group-list
+// To list AAD groups, this code needs to have the clientID
+// of an application with the following permissions:
+// API: Microsoft Graph
+//   Application permissions:
+//      Read all groups
+
+const (
+	clientID     = "0d18f31e-2d82-45cf-8961-dbf2f62dba95" // jminter-groups-test
+	clientSecret = "Updy8eC6NTs9Xs4ltbk158F4BjzqHVYt1Dvj7Tk6pfc="
+	tenantID     = "7a1815fb-63ba-4af7-8f33-e7333e421980"
+)
+
+func fromMSGraphUsers(kubeUser *v1.User, user graphrbac.AADObject, group string) *v1.User {
+	u := kubeUser
+	if u == nil {
+		u = &v1.User{
+			ObjectMeta: meta_v1.ObjectMeta{
+				Labels: map[string]string{
+					"AADObject.ObjectID": *user.ObjectID,
+				},
+			},
+			Groups: []string{group},
+		}
+	}
+	if u.ObjectMeta.Labels == nil {
+		u.ObjectMeta.Labels = map[string]string{}
+	}
+	u.ObjectMeta.Name = *user.UserPrincipalName
+	if user.SignInName != nil {
+		u.ObjectMeta.Labels["AADObject.SignInName"] = *user.SignInName
+	} else {
+		u.ObjectMeta.Labels["AADObject.SignInName"] = ""
+	}
+	u.FullName = *user.DisplayName
+	return u
+}
+
+func fromMSGraphGroup(kubeGroup *v1.Group, group graphrbac.ADGroup, groupName string, msGroupMembers []graphrbac.AADObject) *v1.Group {
+	g := kubeGroup
+	if g == nil {
+		g = &v1.Group{
+			ObjectMeta: meta_v1.ObjectMeta{
+				Name: groupName,
+			},
+			Users: []string{},
+		}
+	}
+	if g.ObjectMeta.Labels == nil {
+		g.ObjectMeta.Labels = map[string]string{}
+	}
+	g.ObjectMeta.Labels["AADObject.ObjectID"] = *group.ObjectID
+	for _, user := range msGroupMembers {
+		g.Users = append(g.Users, *user.DisplayName)
+	}
+	return g
+}
+
+func main() {
+	logrus.SetLevel(logrus.DebugLevel)
+	logrus.SetFormatter(&logrus.TextFormatter{FullTimestamp: true})
+	log := logrus.NewEntry(logrus.StandardLogger())
+
+	gc, err := NewAADGroupsClient()
+	if err != nil {
+		panic(err)
+	}
+	cli, err := openshift.NewAdminClient()
+	if err != nil {
+		panic(err)
+	}
+
+	aadGroups, err := getAllAADGroups(gc)
+	if err != nil {
+		panic(err)
+	}
+
+	groupMap := map[string]string{
+		"Engineering": "osa-customer-admins",
+		"AngusGroup":  "osa-customer-readers",
+	}
+	validGroups := map[string]struct{}{
+		"osa-customer-admins":  struct{}{},
+		"osa-customer-readers": struct{}{},
+	}
+	for _, group := range aadGroups {
+		msGroupMembers, err := getAADGroupMembers(gc, *group.ObjectID)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		kgroupName, ok := groupMap[*group.DisplayName]
+		if !ok {
+			kgroupName = *group.DisplayName

```

Comment:
so this will currently sync non-mapped groups as well - we might want to just do a "continue" here and only sync osa-* mapped groups